README
This is an SX52/XGS emulator/debugger.

Requires: Scons and ncurses

Instructions:
- Load a hex file (assembled SX52 sourcecode) e.g.: ./sxemu tut001.hex

Keys:
TAB - Toggle between RAM and ROM
Page UP/DOWN and Cursor Keys - Browse memory
Space - Execute current command
Return - Quit
b - Set breakpoint
r - Run 
w - Walk (2 commands per second)
R - reset


Changelog
0.03 - 20.08.05

- bugs in the command interpreter were removed


0.04 - 22.08.05

- run/walk mode
- breakpoints
- mnemonics are displayed instead of the raw memory


0.05 - 28.08.05

- several bugfixes
- the emulated program can be restarted


0.06 - 04.09.05

- lots of bugfixes in the command interpreter
  (most of the commands should be error free)
- rtcc is working  
- 'R' resets the emulated program


0.07 - 26.09.05

- some bugfixes (rl,dc flag)
- SRAM access works
- cursor follows pc
- screen layout changed



cpu.h
/*
* =====================================================================================
* 
*        Filename:  cpu.h
* 
*     Description:  
* 
*         Version:  1.0
*         Created:  13.07.2005 23:17:20 CEST
*        Revision:  none
*        Compiler:  gcc
* 
*          Author:  Rainer Blessing 
*           Email:  rainer@theblessing.net
* 
* =====================================================================================
*/
#ifndef CPU
#define CPU
#define USHORT unsigned short

class Cpu
{
public:
	Cpu();
	USHORT* getROM();
	unsigned char* getRAM();
	USHORT* getPC();
	unsigned char* getMode();
	string lookup();
	string lookup(USHORT opc);
	bool execute();
	void reset();
private:
	void sram();
	unsigned char RAM[262];
	USHORT ROM[4096];
	unsigned char SRAM[131072];
	unsigned char* indf;
	unsigned char* rtcc;
	USHORT pc;
	unsigned char* status;
	unsigned char* fsr;
	unsigned char* ra;
	unsigned char* rb;
	unsigned char* rc;
	unsigned char* rd;
	unsigned char* re;
	unsigned char* w;
	USHORT Stack[8];
	USHORT* sp;
	unsigned char mode;
	unsigned char option; //8 bit
	unsigned char Tris[5];
	int cycle;
	int cycle2;
	int prescaler;
};
#endif



sxkey.h
/*
* SxKey.h, part of gSXprog
*
* Copyright 2001-2003 Eric Smith <eric@brouhaha.com>
*
* $Id: SxKey.h,v 1.5 2003/11/18 13:59:42 garlick Exp $
*/

#ifndef SXKEY_H
#define SXKEY_H

#include <string>

#define SX_52_MODE(dev)		(((dev) & 0xfff) == 2)  /* SX48/52 */
#define SX_28_MODE(dev)		(((dev) & 0xfff) != 2)  /* SX18/28 */

#define SXKEY_TYPE_NEW 		0       /* SX-Key Rev. E, Rev. F, SX-Blitz */
#define SXKEY_TYPE_OLD 		1       /* SX-Key Rev. C */

#define SX_MAX_ROM_SIZE 	4096
#define SX_ID_SIZE 		16

/* hex file offsets */
#define ID_LOC 			0x1000
#define FUSE_LOC 		0x1010
#define FUSEX_LOC 		0x1011

/* fusex bits for trimming internal RCA oscillator (modified if --trim) */
#define IRCTRIM0            0x100
#define IRCTRIM1            0x200
#define IRCTRIM2            0x800

/* --clockon limits */
#define MIN_CLOCK   0.4
#define MAX_CLOCK   120.
/* 20 MHz reference clock */
#define REF_CLOCK   20.

#define SXKEY_ECHO_TIMEOUT 	100

#define SXKEY_NEW_TIMEOUT                       (100) /* dflt SX-Key response timeout */
#define SXKEY_NEW_PROG_FUSE_TIMEOUT(n)  (100 + SX_NEW_FUSE_PROG_MS*(n))
#define SXKEY_NEW_PROG_CODE_TIMEOUT(n)  (100 + SX_NEW_CODE_PROG_MS*(n))

/* Programming times */
#define SX_NEW_ERASE_MS                    1000    /* 1...65536 */
#define SX_NEW_FUSE_PROG_MS                 250     /* 1...255 */
#define SX_NEW_CODE_PROG_MS                 100     /* 1...255 */

/* Firmware Revisions */
#define SXKEY_REV_E                     1
#define SXKEY_REV_F                     2
#define SXKEY_BLITZ                     3

/* SX-Key commands */
#define SXKEY_NEW_CMD_END                   0
#define SXKEY_NEW_CMD_ERASE                 1
#define SXKEY_NEW_CMD_PROGRAM               2
#define SXKEY_NEW_CMD_READ                  3
#define SXKEY_NEW_CMD_RESET                 4
#define SXKEY_NEW_CMD_DEBUG_POLL            5
#define SXKEY_NEW_CMD_DEBUG_RECV            6
#define SXKEY_NEW_CMD_DEBUG_SEND            7
#define SXKEY_NEW_CMD_CLOCK_ENABLE          8
#define SXKEY_NEW_CMD_CLOCK_DISABLE         9


#define SXKEY_DEFAULT_DEVICE    "/dev/ttyS0"

#define ARRAY_SIZE 		(SX_MAX_ROM_SIZE + SX_ID_SIZE + 2 + 256)
typedef struct
{
	u16 device;
	u16 rom[ ARRAY_SIZE ];
}
sx_t;

#define SXKEY_T_INIT {              \
                                    .cmd_read = NULL,               \
                                    .cmd_read_fuses = NULL,         \
                                    .cmd_prog = NULL,               \
                                    .cmd_erase = NULL,              \
                                    .cmd_clock_on = NULL,           \
                                    .cmd_clock_off = NULL,          \
                                    .cmd_reset = NULL,              \
                                    .fini = NULL,                   \
                                    .fw_rev = -1,                   \
                                    .fd = -1,                       \
                                    .speed = 0,                     \
                                    .device = SXKEY_DEFAULT_DEVICE, \
                     }
#define speed_t unsigned int

#define ABS(i)  ((i) < 0 ? -(i) : (i))
#define SXKEY_CLK_HZ_TARGET     4000000
#define SXKEY_CLK_HZ(rsp)       (1266035968L/(rsp))

class trim_t
{
public:
	u16 fusex[ 8 ];
	u16 freq[ 8 ];
};


class Fusebits
{
public:
	void clear()
	{
		f_turbo = 1; //sx28
		f_sync = 1;
		f_irc = 1;
		f_div = 0;
		f_ifbd = 1;
		f_xtl = 1; //sx52
		f_cp = 1;
		f_wdte = 0;
		f_osc = 3;
		fx_pins = 1; //sx28
		fx_optionx = 0; //sx28
		fx_sleep = 1; //sx52
		fx_cf = 1;
		fx_bor = 3;
		fx_bp = 3; //sx28
		fx_drt = 1; //sx52
		device = 1; // 0=sx28, 1=sx52
		//u16 deviceString;
	};


	bool f_turbo; //sx28
	bool f_sync;
	bool f_irc;
	char f_div;
	bool f_ifbd;
	bool f_xtl; //sx52
	bool f_cp;
	bool f_wdte;
	char f_osc;

	bool fx_pins; //sx28
	bool fx_optionx; //sx28
	bool fx_sleep; //sx52
	bool fx_cf;
	int fx_bor;
	char fx_bp; //sx28
	char fx_drt; //sx52
	bool device; // 0=sx28, 1=sx52
	u16 deviceString;
};

using std::string;
class SxKey
{
public:
	SxKey();
	SxKey( string device, speed_t bit_rate );
	~SxKey();
	bool sxopen();
	void sxclose();
	bool init();
	Fusebits* getFusebits();
	void saveFusebits();
	char* print();
	void save( char *fn );
	void load( char *fn );
	int  programMemoryProlog( bool trim, bool idskip );
	bool program_20( int i );
	bool programMemoryEpilog( bool idskip );
	int readMemoryProlog();
	bool readMemory( int i );
	bool readMemoryEpilog();
private:
	bool reconnect();
	void unsetkeys();
	void setkeys( u8 c1, u8 c2 );
	bool send_byte( u8 c );
	bool serial_write( u8 c );
	bool serial_read( int ms_timeout, u8 * c );
	u8 encrypt( u8 p );
	u8 decrypt( u8 c );
	void munge_keys();
	bool serial_flush();
	bool send_str( char *s );
	bool expect_str( int ms_timeout, char *s );
	bool get_byte( int ms_timeout, u8 * c );
	bool read_fuses();
	bool get_word( int ms_timeout, u16 * w );
	bool check_error( int ms_timeout );
	bool read_prolog();
	bool read_epilog();
	bool reset();
	void sx52_show_fuse();
	void sx28_show_fuse();

	bool read_id();
	int sx_rom_size();
	void prog_size( int *size, bool idskip );
	bool clock_off();
	bool trim_oscillator( u16 *fusex );
	bool erase();
	bool prog_fusex( u16 fusex, u16 *freq );
	bool prog_prolog();
	bool prog_epilog( u16 *freq );
	bool send_word( u16 w );
	void show_fuse_bits();
	bool prog_verify( bool idskip );
	void sx52_save_fuse();
	void sx28_save_fuse();

	string device;
	int fd;
	speed_t speed;
	bool _crypto_enabled;
	/* values used in stream encryption */
	u8 _crypto_key_1;
	u8 _crypto_key_2;
	u8 fw_rev;
	sx_t sx;
	Fusebits fusebits;
	char *dump;
} ;

#endif

/*
 * em:Local Variables: 
 * em:indent-tabs-mode:nil
 * em:c-basic-offset:4
 * em:End:
 * vi:tabstop=4 shiftwidth=4 expandtab
 */


hex_file.cpp
/*
* hex_file.c, part of gSXprog
*
* Copyright 2001-2003 Eric Smith <eric@brouhaha.com>
*
* $Id: hex_file.c,v 1.5 2003/11/18 13:59:42 garlick Exp $
*/


#include <stdio.h>
#include <termios.h>
#include "hex_file.h"

int HexFile::parse_hex_digit( char **p )
{
	int r = -1;
	if ( ( ( **p ) >= '0' ) && ( ( **p ) <= '9' ) )
		r = **p - '0';
	else if ( ( ( **p ) >= 'A' ) && ( ( **p ) <= 'F' ) )
		r = 10 + **p - 'A';
	else if ( ( ( **p ) >= 'a' ) && ( ( **p ) <= 'f' ) )
		r = 10 + **p - 'a';
	( *p ) ++;
	return ( r );
}

int HexFile::parse_hex_digits( char **p, int digits, u8 * checksum )
{
	int r = 0;
	int d;
	while ( digits-- )
	{
		d = parse_hex_digit( p );
		if ( d < 0 )
			return ( -1 );
		r = ( r << 4 ) + d;
		if ( checksum && !( digits & 1 ) )
			( *checksum ) += ( r & 0xff );
	}
	return ( r );
}

bool HexFile::readFile( char *fn, sx_t * sx )
{
	bool r = 0;
	FILE *f = NULL;
	char *p;
	char buf[ 100 ];
	int i;
	u8 checksum;
	int byte_count;
	int address;
	int record_type;
	int byte;

	f = fopen( fn, "r" );
	if ( !f )
	{
		perror( fn );
		goto error;
	}

	sx->rom[ FUSEX_LOC ] = 0xfff;
	sx->rom[ FUSE_LOC ] = 0xfff;

	for ( i = 0; i < SX_MAX_ROM_SIZE; i++ )
		sx->rom[ i ] = 0xfff;

	for ( i = 0; i < SX_ID_SIZE; i++ )
		sx->rom[ ID_LOC + i ] = 0xfff;

	while ( fgets( &buf[ 0 ], sizeof( buf ), f ) )
	{
		checksum = 0;
		p = &buf[ 0 ];
		while ( ( *p ) && ( ( *p ) != ':' ) )
			p++;
		if ( *p != ':' )
		{
			fprintf( stderr, "warning: spurious line in hex file\n"
			       );
			continue;
		}
		p++;
		byte_count = parse_hex_digits( &p, 2, &checksum );
		if ( byte_count < 0 )
		{
			fprintf( stderr,
			         "error reading hex file: bogus byte count\n"
			       );
			goto error;
		}
		address = parse_hex_digits( &p, 4, &checksum );
		if ( address < 0 )
		{
			fprintf( stderr, "error reading hex file: bogus address\n"
			       );
			goto error;
		}
		record_type = parse_hex_digits( &p, 2, &checksum );
		if ( record_type < 0 )
		{
			fprintf( stderr,
			         "error reading hex file: bogus record type\n"
			       );
			goto error;
		}
		if ( record_type > 1 )
		{
			/* Ignore expected extra gpasm data, but warn about others */
			if ( ( record_type == 0x04 ) && ( byte_count == 0x02 ) &&
			        ( address == 0x0000 ) )
				continue;
			fprintf( stderr,
			         "warning: unknown record type %02x in hex file\n",
			         record_type );
			continue;
		}
		if ( ( record_type == 1 ) && ( byte_count != 0 ) )
		{
			fprintf( stderr,
			         "error reading hex file: end record with data bytes\n"
			       );
			goto error;
		}
		for ( i = 0; i < byte_count; i++ )
		{
			byte = parse_hex_digits( &p, 2, &checksum );
			if ( byte < 0 )
			{
				fprintf( stderr,
				         "error reading hex file: bogus data byte for addr %04x\n",
				         address + i );
				goto error;
			}
			if ( ( address + i ) > ( ARRAY_SIZE * 2 ) )
			{
				fprintf( stderr,
				         "error reading hex file: address %04x out of boundsx\n",
				         address + i );
				goto error;
			}
			if ( ( address + i ) & 1 )
				sx->rom[ ( address + i ) / 2 ] =
				    ( sx->rom[ ( address + i ) / 2 ] & 0x00ff )
				    + ( byte << 8 );
			else
				sx->rom[ ( address + i ) / 2 ] =
				    ( sx->rom[ ( address + i ) / 2 ] & 0xff00 )
				    + ( byte );
		}
		byte = parse_hex_digits( &p, 2, &checksum );
		if ( byte < 0 )
		{
			fprintf( stderr, "error reading hex file: bogus checksum\n" );
			goto error;
		}
		if ( checksum != 0 )
		{
			fprintf( stderr,
			         "error reading hex file: incorrect checksum\n" );
			goto error;
		}
		if ( byte_count == 0 )
			break;              /* logical end of file */
	}

	r = 1;

error:
	if ( f )
		fclose( f );
	return ( r );
}


void HexFile::writeByte( FILE * f, u8 byte, u8 * checksum )
{
	fprintf( f, "%02X", byte );
	if ( checksum )
		*checksum += byte;
}

bool HexFile::writeFile( char *fn, sx_t * sx )
{
	FILE *f = NULL;
	int r = 0;
	int i, j;
	u8 checksum;

	f = fopen( fn, "w" );
	if ( !f )
	{
		perror( fn );
		goto error;
	}
	for ( i = 0; i < SX_MAX_ROM_SIZE; i += 8 )
	{
		checksum = 0;
		fprintf( f, ":" );
		writeByte( f, 0x10, &checksum );     /* byte count */
		writeByte( f, ( i * 2 ) >> 8, &checksum );     /* address high */
		writeByte( f, ( i * 2 ) & 0xff, &checksum );   /* address low */
		writeByte( f, 0x00, &checksum );     /* record type */
		for ( j = i; j < ( i + 8 ); j++ )
		{
			writeByte( f, sx->rom[ j ] & 0xff, &checksum );
			writeByte( f, sx->rom[ j ] >> 8, &checksum );
		}
		writeByte( f, 256 - checksum, NULL );
		fprintf( f, "\n" );
	}
	for ( i = ID_LOC; i < ID_LOC + SX_ID_SIZE; i += 8 )
	{
		checksum = 0;
		fprintf( f, ":" );
		writeByte( f, 0x10, &checksum );     /* byte count */
		writeByte( f, ( i * 2 ) >> 8, &checksum );     /* address high */
		writeByte( f, ( i * 2 ) & 0xff, &checksum );   /* address low */
		writeByte( f, 0x00, &checksum );     /* record type */
		for ( j = i; j < ( i + 8 ); j++ )
		{
			writeByte( f, sx->rom[ j ] & 0xff, &checksum );
			writeByte( f, sx->rom[ j ] >> 8, &checksum );
		}
		writeByte( f, 256 - checksum, NULL );
		fprintf( f, "\n" );
	}

	checksum = 0;
	fprintf( f, ":" );
	writeByte( f, 0x04, &checksum ); /* byte count */
	writeByte( f, ( FUSE_LOC * 2 ) >> 8, &checksum );  /* address high */
	writeByte( f, ( FUSE_LOC * 2 ) & 0xff, &checksum );        /* address low */
	writeByte( f, 0x00, &checksum ); /* record type */
	for ( j = FUSE_LOC; j <= FUSEX_LOC; j++ )
	{
		writeByte( f, sx->rom[ j ] & 0xff, &checksum );
		writeByte( f, sx->rom[ j ] >> 8, &checksum );
	}
	writeByte( f, 256 - checksum, NULL );
	fprintf( f, "\n" );

	checksum = 0;
	fprintf( f, ":" );
	writeByte( f, 0x00, &checksum ); /* byte count */
	writeByte( f, 0x00, &checksum ); /* address high */
	writeByte( f, 0x00, &checksum ); /* address low */
	writeByte( f, 0x01, &checksum ); /* record type */
	writeByte( f, 256 - checksum, NULL );
	fprintf( f, "\n" );

	r = 1;

error:
	if ( f )
		fclose( f );
	return ( r );
}

/*
 * em:Local Variables: 
 * em:indent-tabs-mode:nil
 * em:c-basic-offset:4
 * em:End:
 * vi:tabstop=4 shiftwidth=4 expandtab
 */


sxemu.cpp
using namespace std;
#include <unistd.h>
#include 	<pthread.h>
#include	<iostream>
#include	<iomanip>
#include 	<curses.h>
#include	"filehandler.h"
#include 	<string>
#include	"cpu.h"

/*-----------------------------------------------------------------------------
 *  function main
 *-----------------------------------------------------------------------------*/

void dumpMem(char *dump, USHORT* mem, USHORT size);
void dumpRAM(char *dump, unsigned char* mem, USHORT size);
void screen();
bool keys();
void run();

Cpu cpu;
WINDOW *mainwin, *romwin, *ramwin, *sidewin;
string impl = "";
unsigned char* romptr;
unsigned char* ramptr;
USHORT* pc;
unsigned char memrow[46];
unsigned char ramrow[53];
int romr, ramr;
char awin; // tab changes window focus
unsigned char* RAM;
USHORT* ROM;
int row, col;
int cursorrow;
bool breakpoint[4096];
bool running = false;
pthread_t walkthread;

void *run(void *Arg)
{
	while (running == true)
	{
		romr = 35 * (*pc);
		cursorrow = romr / 35;
		if (cpu.execute() == false )
		{
			running = false;
			break;
		}
		if (*(char*)Arg == 'w' )usleep(500000);
		if (breakpoint[*pc] == true)
		{
			running = false;
			break;
		}
	}
	return NULL;
}

int
main (int argc, char *argv[])
{
	FileHandler fh;

	RAM = cpu.getRAM();
	ROM = cpu.getROM();
	if (argc < 2)
	{
		printf("Usage: sxemu <hexfile>\n");
		printf("e.g. \"./sxemu file.hex\"\n");
		return EXIT_FAILURE;
	}
	fh.readFile (argv[1], ROM);
	if ((mainwin = initscr ()) == NULL)
	{
		fprintf (stderr, "Error initialising ncurses.\n");
		exit (EXIT_FAILURE);
	}

	noecho();
	start_color();
	curs_set(0);
	nodelay(stdscr, true);
	keypad (mainwin, TRUE);
	getmaxyx(stdscr, row, col);
	romwin = subwin (mainwin, row, 27, 0, 0);
	ramwin = subwin (mainwin, row / 2, 54, row / 2 + 1, 27);
	sidewin = subwin (mainwin, row / 2, 54, 0, 27);

	cursorrow = 0;
	ramr = romr = 0;
	//romptr = new char[(4096 * 2 / 8) * (32 + 8 + 5) + 1];
	romptr = new unsigned char[4096 * 35];
	dumpMem((char*)romptr, ROM, 4096);
	ramptr = new unsigned char[(262 * 2 / 8) * (48 + 5 + 8) + 1];
	memrow[45] = '\0';
	ramrow[53] = '\0';
	pc = cpu.getPC();
	init_pair(1, COLOR_BLACK, COLOR_GREEN);
	init_pair(2, COLOR_RED, COLOR_BLACK);

	string impl = "";
	while (1)
	{
		screen();
		if (keys() == false)break;
	}
	delete[] ramptr;
	delete[] romptr;

	delwin (romwin);
	delwin (ramwin);
	delwin (sidewin);
	delwin (mainwin);
	endwin();
	running = false;
	if (walkthread != NULL)pthread_join( walkthread, NULL);
	return EXIT_SUCCESS;
}				// ----------  end of function main  ----------

void screen()
{
	int row, col;
	impl = cpu.lookup();
	getmaxyx(stdscr, row, col);
	erase();
	mvwprintw(ramwin, 1, 1, "     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f");
	dumpRAM((char*)ramptr, cpu.getRAM(), 262);
	for (int j = 0;j < row - 2;j++)
	{
		memcpy(memrow, romptr + romr + j*35, 35);
		mvwprintw (romwin, 1 + j, 1, "%s", memrow);
		if (romr / 35 + j == cursorrow)mvchgat(1 + j, 6 , 20, A_REVERSE, 0, NULL);
		if (romr / 35 + j == *pc)
		{
			mvchgat(1 + j, 0 , 4, A_REVERSE, 1, NULL);
		}
		if (breakpoint[romr / 35 + j] == true)
		{
			mvchgat(1 + j, 5 , 1, A_REVERSE, 2, NULL);
		}
	}
	for (int j = 0;j < row / 2 - 3;j++)
	{
		memcpy(ramrow, ramptr + ramr + j*53, 52);
		mvwprintw (ramwin, 2 + j, 1, "%s", ramrow);
	}
	if (awin == 0)
	{
		wattron(romwin, A_BOLD);
	}
	else
	{
		wattron(ramwin, A_BOLD);
	}
	box (romwin, 0, 0);
	box (ramwin, 0, 0);
	box (sidewin, 0, 0);
	wattroff(romwin, A_BOLD);
	wattroff(ramwin, A_BOLD);
	mvwprintw(sidewin, 1, 1, "INDF %02X", RAM[RAM[4]]);
	mvwprintw(sidewin, 2, 1, "RTCC %02X", RAM[1]);
	mvwprintw(sidewin, 3, 1, "PC   %03X", *pc);
	mvwprintw(sidewin, 4, 1, "ST   %02X", RAM[3]);
	mvwprintw(sidewin, 5, 1, "FSR  %02X", RAM[4]);
	mvwprintw(sidewin, 6, 1, "RA   %02X", RAM[5]);
	mvwprintw(sidewin, 7, 1, "RB   %02X", RAM[6]);
	mvwprintw(sidewin, 8, 1, "RC   %02X", RAM[7]);
	mvwprintw(sidewin, 9, 1, "RD   %02X", RAM[8]);
	mvwprintw(sidewin, 10, 1, "RE   %02X", RAM[9]);
	mvwprintw(sidewin, 11, 1, "W    %02X", RAM[261]);
	mvwprintw(sidewin, 1, 20, "Mode %02X", *cpu.getMode());
	mvwprintw(sidewin, 2, 20, "OPC  %03X", ROM[*pc]);
	int j = 0;
	for (int i = 0;i < 12;i++)
	{
		int k = ROM[*pc] >> i & 1;
		mvwprintw(sidewin, 2, 30 + 15 - (j++), "%d", k);
		if ((i + 1) % 4 == 0)
			mvwprintw(sidewin, 13, 1 + 15 - (j++), " ");
		mvwprintw(sidewin, 15, 1, "%s", impl.c_str());
	}
	refresh();
}

bool keys()
{
	int ch = getch();
	switch (ch)
	{
			case KEY_UP:
			if (awin == 0)
			{
				if (cursorrow > 0)
				{
					cursorrow--;
					if (cursorrow*35 < romr)romr -= 35;
				}
			}
			else ramr = ramr - 53;
			break;
			case KEY_PPAGE:
			if (awin == 0)
			{
				romr = romr - (row - 2) * 35;
				cursorrow = romr / 35;
			}
			else
				ramr = ramr - (row / 2 - 2) * 53;

			break;
			case KEY_DOWN:
			if (awin == 0)
			{
				if (cursorrow < 4095)
				{
					cursorrow++;
					if (cursorrow*35 >= romr + (row - 2) * 35)romr += 35;
				}
			}
			else
				ramr = ramr + 53;
			break;
			case KEY_NPAGE:
			if (awin == 0)
			{
				romr = romr + (row - 2) * 35;
				cursorrow = romr / 35;
			}
			else ramr = ramr + (row / 2 - 2) * 53;
			break;
			case 10:      // enter
			ch = 0;
			break;
			case 32:      // space
			if (running == false)
			{
				if (cpu.lookup() == "END")
				{
					cpu.reset();
				}
				cpu.execute();
				romr = 35 * (*pc);
				cursorrow = romr / 35;
			}
			break;
			case 9:      //tab
			if (awin == 0)awin = 1;
			else awin = 0;
			break;
			case 'r':
			if (cpu.lookup() == "END")
			{
				cpu.reset();
			}
			if (running == false)
			{
				running = true;
				char m = 'r';
				pthread_create( &walkthread, NULL, &run, &m);
			}
			else running = false;
			break;
			case 'R':
			running = false;
			cpu.reset();
			break;
			case 'w':
			if (cpu.lookup() == "END")
			{
				cpu.reset();
			}
			if (running == false)
			{
				running = true;
				char m = 'w';
				pthread_create( &walkthread, NULL, &run, &m);
			}
			else running = false;
			break;
			case 'b':
			if (breakpoint[cursorrow] != true)
				breakpoint[cursorrow] = true;
			else
				breakpoint[cursorrow] = false;
			break;
	}
	if (ch == 0) return false;
	if (romr < 0)
	{
		romr = 0;
		cursorrow = 0;
	}
	if (ramr < 0)ramr = 0;
	if ((romr / 35 + row - 2) > 4096)romr = (4096 - row + 2) * 35;
	if ((ramr / 53 + row / 2 - 2) > 32)ramr = (32 - row / 2 + 2) * 53;
	return true;
}

void dumpMem(char *dump, USHORT* mem, USHORT size)
{
	int i, j, k;
	k = 0;
	j = 35;
	for (i = 0; i < size; i ++)
	{
		sprintf (dump + k, "%03X: ", i);
		k += 5;
		string cmd = cpu.lookup(mem[i] & 0xfff);
		sprintf (dump + k, "%s", cmd.c_str() );
		k += cmd.length();
		sprintf (dump + j, "\n");
		k = j;
		j += 35;
	}
}

void dumpRAM(char *dump, unsigned char* mem, USHORT size)
{
	int i, j, k;
	k = 0;

	for (i = 0; i < size; i += 16)
	{
		sprintf (dump + k, "%03X: ", i);
		k += 5;
		for (j = i; j < (i + 16); j++)
		{
			sprintf (dump + k, "%02X ", mem[j] & 0xff);
			k += 3;
		}
		sprintf (dump + k - 1, "\n");
	}
}
